File: learning-spark-exercise/src/main/scala/spark/scala/BasicAvg.scala
package spark.scala

import org.apache.spark._
import org.apache.spark.rdd.RDD

object BasicAvg {

  def main(args: Array[String]): Unit = {
    val master = args.length match {
      case x: Int if x > 0 => args(0)
      case _               => "local"
    }

    val sc = new SparkContext(master, "BasicAvg", System.getenv("SPARK_HOME"))

    val input = sc.parallelize(List(1, 2, 3, 4))

    val result = input.aggregate((0,0))(
      (acc,value) =>(acc._1 + value,acc._2+1),
      (acc1,acc2)=>(acc1._1+acc2._1,acc2._2+acc2._2)
    )
    
    val avg = result._1/result._2.toDouble
  }

}

File: learning-spark-exercise/src/main/java/spark/java/BasicAvg.java
package spark.java;

import java.io.Serializable;
import java.util.Arrays;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;

public class BasicAvg {

	public static class AvgCount implements Serializable {
		public AvgCount(int total, int num) {
			total_ = total;
			num_ = num;
		}

		public int total_;
		public int num_;

		public float avg() {
			return total_ / (float) num_;
		}
	}
	
	public static Function2<AvgCount,Integer,AvgCount> addAndCount = new Function2<AvgCount,Integer,AvgCount>(){

		@Override
		public AvgCount call(AvgCount a, Integer x) throws Exception {
			a.total_ += x;
			a.num_ += 1;
			return a;
		}};
		
	public static Function2<AvgCount,AvgCount,AvgCount> combine =  new Function2<AvgCount,AvgCount,AvgCount>(){

		@Override
		public AvgCount call(AvgCount a, AvgCount b) throws Exception {
			a.total_ += b.total_;
			a.num_ += b.num_;		
			return a;
		}};

	public static void main(String[] args) throws Exception {
		String master;
		if (args.length > 0) {
			master = args[0];
		} else {
			master = "local";
		}
		
		JavaSparkContext  sc = new JavaSparkContext( master, "basicavg", System.getenv("SPARK_HOME"),System.getenv("JARS"));
		
		JavaRDD<Integer> rdd = sc.parallelize(Arrays.asList(1,2,3,4));
		
		AvgCount initial = new AvgCount(0, 0);
		
		AvgCount result = rdd.aggregate(initial, addAndCount, combine);
		System.out.println(result.avg());
		
	}
}
